File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/CollectionOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.ConfigHelper;
import io.github.chindeaone.collectiontracker.config.categories.Bazaar;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.tracker.collection.TrackingHandler;
import io.github.chindeaone.collectiontracker.utils.HypixelUtils;
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils;
import io.github.chindeaone.collectiontracker.utils.rendering.TextUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.ChatScreen;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class CollectionOverlay implements AbstractOverlay{

    public static volatile boolean trackingDirty = false;
    private final Position position = ConfigAccess.getTrackingPosition();
    public final List<String> overlayLines = new ArrayList<>();
    public final List<String> extraOverlayLines = new ArrayList<>();
    private boolean renderingAllowed  = true;

    @Override
    public String overlayLabel() {
        return "Collection Tracker";
    }

    @Override public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return TrackingHandler.isTracking && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled() || !trackingDirty) return;

        List<String> mainLines = getCollectionLines();
        List<String> extraLines = getCollectionExtraLines();

        if (mainLines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
            RenderUtils.renderTrackingStringsWithColor(context, mainLines, extraLines, ConfigAccess.isOverlayTextColorEnabled())
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getCollectionLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }

    @Override
    public List<String> getLines() {
        if (!isEnabled()) return Collections.emptyList();

        List<String> combinedLines = new ArrayList<>(getCollectionLines());
        if (ConfigAccess.isShowExtraStats()) {
            // append extra lines to create a single list
            combinedLines.add(""); // add separator line
            combinedLines.addAll(getCollectionExtraLines());
        }

        return combinedLines;
    }

    @Override
    public void handleLineAction(String line) {
        switch (line) {
            case "§e[Bazaar Prices]" -> ConfigHelper.setBazaar(true);
            case "§e[NPC Prices]" -> ConfigHelper.setBazaar(false);
            case "§e[Extra Stats]" -> ConfigHelper.setShowExtraStats(!ConfigAccess.isShowExtraStats());
        }
        if (line.contains(ConfigAccess.getGemstoneVariant().toString())) {
            cycleGemstoneVariant();
        }
        if (line.contains("version")) {
            changeEnchantedType();
        }
        if (line.contains("Instant")) {
            changeBazaarPriceType();
        }
    }

    @Override
    public boolean isHovered(double mouseX, double mouseY) {
        if (!isEnabled()) return false;

        List<String> mainLines = getCollectionLines();
        if (mainLines.isEmpty()) return false;

        Position pos = this.position();
        float scale = pos.getScale();
        Font fr = Minecraft.getInstance().font;

        int totalLines = mainLines.size();

        if (ConfigAccess.isShowExtraStats()) {
            List<String> extraLines = getCollectionExtraLines();
            if (!extraLines.isEmpty()) {
                totalLines += 1; // add separator line
                totalLines += extraLines.size();
            }
        }

        int height = (int) (fr.lineHeight * totalLines * scale);
        int width = (int) (pos.getWidth() * scale);
        int x = pos.getX();
        int y = pos.getY();

        return mouseX >= x && mouseX <= x + width && mouseY >= y && mouseY <= y + height;
    }

    private @NotNull List<String> getCollectionLines() {
        TextUtils.updateTrackingLines(overlayLines);
        if (overlayLines.isEmpty()) return overlayLines;
        List<String> lines = new ArrayList<>(overlayLines);
        lines.add("Uptime: " + TrackingHandler.getUptime());
        if (!ConfigAccess.isShowExtraStats()) {
            if (Minecraft.getInstance().screen instanceof ChatScreen) {
                TextUtils.addToggleableSettingsLines(lines);
            }
        }
        return lines;
    }

    private @NotNull List<String> getCollectionExtraLines() {
        if (!ConfigAccess.isShowExtraStats()) return Collections.emptyList();
        TextUtils.updateTrackingExtraLines(extraOverlayLines);
        if (Minecraft.getInstance().screen instanceof ChatScreen) {
            TextUtils.addToggleableSettingsLines(extraOverlayLines);
        }
        return extraOverlayLines;
    }

    private void cycleGemstoneVariant() {
        Bazaar.GemstoneVariant[] variants = Bazaar.GemstoneVariant.values();
        Bazaar.GemstoneVariant current = ConfigAccess.getGemstoneVariant();
        int nextOrdinal = (current.ordinal() + 1) % variants.length;
        ConfigHelper.setGemstoneVariant(variants[nextOrdinal]);
    }

    private void changeEnchantedType() {
        ConfigHelper.setBazaarType(ConfigAccess.getBazaarType() == Bazaar.BazaarType.ENCHANTED_VERSION ?
                Bazaar.BazaarType.SUPER_ENCHANTED_VERSION : Bazaar.BazaarType.ENCHANTED_VERSION);
    }

    private void changeBazaarPriceType() {
        ConfigHelper.changeBazaarPrice(ConfigAccess.getBazaarPriceType() == Bazaar.BazaarPriceType.INSTANT_BUY ?
                Bazaar.BazaarPriceType.INSTANT_SELL : Bazaar.BazaarPriceType.INSTANT_BUY);
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/CollectionOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.ModLoader
import io.github.chindeaone.collectiontracker.config.ConfigHelper
import io.github.chindeaone.collectiontracker.config.ConfigHelper.changeBazaarPrice
import io.github.chindeaone.collectiontracker.config.ConfigHelper.setBazaar
import io.github.chindeaone.collectiontracker.config.ConfigHelper.setBazaarType
import io.github.chindeaone.collectiontracker.config.ConfigHelper.setShowExtraStats
import io.github.chindeaone.collectiontracker.config.bazaarConfig
import io.github.chindeaone.collectiontracker.config.bazaarPriceType
import io.github.chindeaone.collectiontracker.config.bazaarType
import io.github.chindeaone.collectiontracker.config.categories.Bazaar
import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.config.gemstoneVariant
import io.github.chindeaone.collectiontracker.config.showExtraStats
import io.github.chindeaone.collectiontracker.config.trackingPosition
import io.github.chindeaone.collectiontracker.tracker.collection.TrackingHandler
import io.github.chindeaone.collectiontracker.tracker.collection.TrackingRates
import io.github.chindeaone.collectiontracker.utils.MinecraftUtils
import io.github.chindeaone.collectiontracker.utils.parser.CollectionParser
import io.github.chindeaone.collectiontracker.utils.render.RenderUtils
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.screens.ChatScreen
import kotlin.concurrent.Volatile
import kotlin.math.roundToInt

class CollectionOverlay : AbstractOverlay() {
    private var cachedLines: List<String> = emptyList()

    override val overlayLabel: String = "Collection Tracker"

    override val position: Position get() = trackingPosition

    override val isEnabled: Boolean get() = TrackingHandler.isTracking

    override fun render(context: GuiGraphicsExtractor) {
        if (!isEnabled || !trackingDirty) return

        val lines = lines
        if (lines.isEmpty()) return

        RenderUtils.drawOverlayFrame(context, position) { RenderUtils.renderTrackingStringsWithColor(context, lines) }
    }

    override fun updateDimensions() {
        if (!isEnabled || !trackingDirty) return
        updateLinesIfNeeded()

        super.updateDimensions()
    }

    override val lines: List<String>
        get() {
            updateLinesIfNeeded()
            return cachedLines
        }

    private fun updateLinesIfNeeded() {
        if (!isEnabled || !trackingDirty) {
            cachedLines = emptyList()
            return
        }

        if (ModLoader.clientTicks % 5L != 0L) return

        TrackingRates.updateRates()

        val uptime = TrackingHandler.uptime
        val isChatOpened = MinecraftUtils.screen is ChatScreen
        val showExtra = showExtraStats

        val main = mutableListOf<String>()
        CollectionParser.updateTrackingLines(main)
        if (main.isNotEmpty()) {
            main.add("Uptime: $uptime")
            if (!showExtra && isChatOpened) CollectionParser.addToggleableSettingsLines(main)
        }

        val extra = mutableListOf<String>()
        if (showExtra) {
            CollectionParser.updateTrackingExtraLines(extra)
            CollectionParser.addToggleableSettingsLines(extra)
        }

        cachedLines = buildList {
            addAll(main)
            if (showExtra) {
                add("") // add separator line
                addAll(extra)
            }
        }
    }

    override fun handleLineAction(line: String) {
        when {
            line.contains("Prices") -> setBazaar(!bazaarConfig.useBazaar)
            line.contains("Extra") -> setShowExtraStats(!showExtraStats)
            line.contains(gemstoneVariant.toString()) -> cycleGemstoneVariant()
            line.contains("version") -> changeEnchantedType()
            line.contains("Instant") -> changeBazaarPriceType()
        }
    }

    override fun isHovered(mouseX: Double, mouseY: Double): Boolean {
        if (!isEnabled) return false

        updateDimensions()

        val position = position

        val padding = 8

        val x = position.x
        val y = position.y
        val scale = position.scale

        val width = ((position.width + padding * 2) * scale).roundToInt()
        val height = ((position.height + padding * 2) * scale).roundToInt()

        val x1 = (x - padding * scale).toDouble()
        val y1 = (y - padding * scale).toDouble()
        val x2 = x1 + width
        val y2 = y1 + height

        return mouseX in x1..x2 && mouseY >= y1 && mouseY <= y2
    }

    private fun cycleGemstoneVariant() {
        val variants: Array<Bazaar.GemstoneVariant> = Bazaar.GemstoneVariant.entries.toTypedArray()
        val current = gemstoneVariant
        val nextOrdinal = (current.ordinal + 1) % variants.size
        ConfigHelper.setGemstoneVariant(variants[nextOrdinal])
    }

    private fun changeEnchantedType() {
        setBazaarType(if (bazaarType == Bazaar.BazaarType.ENCHANTED_VERSION) Bazaar.BazaarType.SUPER_ENCHANTED_VERSION else Bazaar.BazaarType.ENCHANTED_VERSION)
    }

    private fun changeBazaarPriceType() {
        changeBazaarPrice(if (bazaarPriceType == Bazaar.BazaarPriceType.INSTANT_BUY) Bazaar.BazaarPriceType.INSTANT_SELL else Bazaar.BazaarPriceType.INSTANT_BUY)
    }

    companion object {
        @Volatile
        var trackingDirty: Boolean = false
    }
}
